lib/Runtime/Library/JSON.cpp

```diff
@@ -688,55 +689,45 @@ namespace JSON
+                            while ((propertyName = enumerator.MoveAndGetNext(id)) != NULL)
                             {
                                 // if unsuccessful get propertyId from the string
-                                scriptContext->GetOrAddPropertyRecord(propertyName->GetString(), propertyName->GetLength(), &propRecord);
+                                scriptContext->GetOrAddPropertyRecord(propertyName, &propRecord);
+                                if (!propRecord->IsNumeric()) break;
                                 id = propRecord->GetPropertyId();
                                 StringifyMemberObject(propertyName, id, value, (Js::ConcatStringBuilder*)result, indentString, memberSeparator, isFirstMember, isEmpty);
-                                index++;
                             }
                         }
 
                         // loop over the named properties
                         for (uint32 i = 0; i < propertyCount; i++)
                         {
                             id = typeHandler->GetPropertyId(scriptContext, (Js::PropertyId)i);
-
-                            if (id < Js::InternalPropertyIds::Count)
+                            if (id == Js::Constants::NoProperty)
                             {
-                                // if the property ID is internal, we cannot assume dynamicObject->getSlot(i) is a var later
-                                previousId = id;
                                 continue;
                             }
-                            else if (id == Js::Constants::NoProperty)
-                            {
-                                if ((propertyName = enumerator.MoveAndGetNext(previousId)) == NULL) break;
-
-                                scriptContext->GetOrAddPropertyRecord(propertyName->GetString(), propertyName->GetLength(), &propRecord);
-                                id = propRecord->GetPropertyId();
-                            }
                             else
                             {
-                                propRecord = scriptContext->GetPropertyName(id);
-                                propertyName = scriptContext->GetPropertyString(id);
+                                if (!object->IsEnumerable(id)) continue;
+                                Js::PropertyString * propertyString = scriptContext->GetPropertyString(id);
+                                propRecord = propertyString->GetPropertyRecord();
+                                propertyName = (Js::JavascriptString*) propertyString;
                             }
 
-                            previousId = id;
                             if (!propRecord->IsSymbol())
                             {
-                                Js::Var property = dynamicObject->GetSlot(i);
+                                PropertyIndex index = typeHandler->GetPropertyIndex(propRecord);
+                                Js::Var property = index != Constants::NoSlot ?
```

Comment:
How does this work with getters? I believe that `GetPropertyIndex` only gets the `Data` property of the `DictionaryPropertyDescriptor` in that case, which would be `NoSlots` if it has an accessor.

Reply:
That's why `index != Constants::NoSlot`